### PR TITLE
TRAIN-294: get income and expense by userid

### PR DESCRIPTION
**Pull request by Stanley-Maraba**

### CODE REVIEW

File: src/main/java/com/budgetapp/controllers/ExpensesController.java [hunk 1]
```diff
@@ -42,4 +44,14 @@ public ResponseEntity<Expense> saveExpenses(@RequestBody final Expense expense)
         final var createExpenses = expensesService.addExpenses(expense);
         return ResponseEntity.status(HttpStatus.CREATED).body(createExpenses);
     }
+
+    @GetMapping(path = "/{user_id}", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/controllers/ExpensesController.java]: This doesn't follow Java naming conventions

```suggestion
    @GetMapping(path = "/{userId}", produces = MediaType.APPLICATION_JSON_VALUE)
```

File: src/main/java/com/budgetapp/controllers/IncomeController.java [hunk 2]
```diff
@@ -49,4 +50,14 @@ private ResponseEntity<List<Income>> findAllIncomes() {
         final var findIncomes = incomeService.findAllIncomes();
         return ResponseEntity.ok(findIncomes);
     }
+
+    @GetMapping(path = "/{user_id}", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/controllers/IncomeController.java]:
```suggestion
    @GetMapping(path = "/{userId}", produces = MediaType.APPLICATION_JSON_VALUE)
```

File: src/main/java/com/budgetapp/repositories/ExpensesRepository.java [hunk 3]
```diff
@@ -3,5 +3,9 @@
 import com.budgetapp.entities.Expense;
 import org.springframework.data.jpa.repository.JpaRepository;
 
+import java.util.List;
+import java.util.Optional;
+
 public interface ExpensesRepository extends JpaRepository<Expense, Long> {
+    Optional<List<Expense>> findByUserId(final Long user_id);
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/repositories/ExpensesRepository.java]: Please fix everywhere that you have named this `user_id`

**Review by JOlsen-Neslo — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 4]
```diff
@@ -51,6 +51,11 @@
             <groupId>org.postgresql</groupId>
             <artifactId>postgresql</artifactId>
         </dependency>
+        <dependency>
```
Comment (JOlsen-Neslo) [inline pom.xml]: Why was this added? If you are using spring data this isn't needed

**Review by JOlsen-Neslo — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 4 above)
Comment (Stanley-Maraba) [reply pom.xml]: wrongly added indeed not needed

fixed

File: src/main/java/com/budgetapp/controllers/ExpensesController.java [hunk 5]
```diff
@@ -42,4 +43,14 @@ public ResponseEntity<Expense> saveExpenses(@RequestBody final Expense expense)
         final var createExpenses = expensesService.addExpenses(expense);
         return ResponseEntity.status(HttpStatus.CREATED).body(createExpenses);
     }
+
+    @GetMapping(path = "/{userId}", produces = MediaType.APPLICATION_JSON_VALUE)
+    private ResponseEntity<List<Expense>> findIncomesById(@PathVariable final Long userId) {
+        try {
+            List<Expense> expense = expensesService.findExpensesByUserId(userId);
+            return ResponseEntity.ok(expense);
+        } catch (EntityNotFoundException e) {
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/controllers/ExpensesController.java]: Following best practices for RESTful API design, if there are none found, you should rather return an empty list with a 200 response rather than a not found

File: src/main/java/com/budgetapp/repositories/IncomeRepository.java [hunk 6]
```diff
@@ -3,5 +3,9 @@
 import com.budgetapp.entities.Income;
 import org.springframework.data.jpa.repository.JpaRepository;
 
+import java.util.List;
+import java.util.Optional;
+
 public interface IncomeRepository extends JpaRepository<Income, Long> {
+    Optional<List<Income>> findByUserId(final Long userId);
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/repositories/IncomeRepository.java]: This shouldn't be an optional

```suggestion
    List<Income> findAllByUserId(final Long userId);
```

File: src/main/java/com/budgetapp/services/ExpensesService.java [hunk 7]
```diff
@@ -26,4 +27,8 @@ public void deleteExpensesById(final Long id) {
     public Expense addExpenses(final Expense expense) {
         return expensesRepository.save(expense);
     }
+
+    public List<Expense> findExpensesByUserId(final Long userId) {
+        return expensesRepository.findByUserId(userId).orElseThrow(EntityNotFoundException::new);
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/services/ExpensesService.java]: Don't throw an exception

**Review by JOlsen-Neslo — CHANGES_REQUESTED** [review]

File: src/main/java/com/budgetapp/controllers/ExpensesController.java (hunk 5 above)
Comment (Stanley-Maraba) [reply src/main/java/com/budgetapp/controllers/ExpensesController.java]: isn't a 200 response meant for successful request?

Comment (colsen-neslo) [reply src/main/java/com/budgetapp/controllers/ExpensesController.java]: Yes they are but technically it is a successful request if the service can't find any expenses yet if the user has none.

File: src/main/java/com/budgetapp/repositories/ExpensesRepository.java [hunk 8]
```diff
@@ -3,5 +3,9 @@
 import com.budgetapp.entities.Expense;
 import org.springframework.data.jpa.repository.JpaRepository;
 
+import java.util.List;
+import java.util.Optional;
+
 public interface ExpensesRepository extends JpaRepository<Expense, Long> {
+    Optional<List<Expense>> findByUserId(final Long userId);
```
Comment (colsen-neslo) [inline src/main/java/com/budgetapp/repositories/ExpensesRepository.java]:
```suggestion
    List<Expense> findAllByUserId(final Long userId);
```

File: src/main/java/com/budgetapp/repositories/IncomeRepository.java (hunk 6 above)
Comment (colsen-neslo) [reply src/main/java/com/budgetapp/repositories/IncomeRepository.java]: This is not resolved.

File: src/main/java/com/budgetapp/services/ExpensesService.java (hunk 7 above)
Comment (colsen-neslo) [reply src/main/java/com/budgetapp/services/ExpensesService.java]: This is not resolved.

File: src/main/java/com/budgetapp/services/IncomeService.java [hunk 9]
```diff
@@ -23,7 +24,11 @@ public void deleteIncome(final Long id) {
         incomeRepository.deleteById(id);
     }
 
-    public List<Income> findAllIncomes(){
+    public List<Income> findAllIncomes() {
         return incomeRepository.findAll();
     }
+
+    public List<Income> findIncomesByUserId(final Long userId) {
+        return incomeRepository.findByUserId(userId).orElseThrow(EntityNotFoundException::new);
```
Comment (colsen-neslo) [inline src/main/java/com/budgetapp/services/IncomeService.java]: Don't throw an exception.

File: src/main/java/com/budgetapp/controllers/IncomeController.java [hunk 10]
```diff
@@ -49,4 +50,14 @@ private ResponseEntity<List<Income>> findAllIncomes() {
...
     }
+
+    @GetMapping(path = "/{userId}", produces = MediaType.APPLICATION_JSON_VALUE)
+    private ResponseEntity<List<Income>> findIncomesById(@PathVariable final Long userId) {
+        try {
+            List<Income> income = incomeService.findIncomesByUserId(userId);
+            return ResponseEntity.ok(income);
+        } catch (EntityNotFoundException e) {
+            return ResponseEntity.notFound().build();
+        }
```
Comment (colsen-neslo) [inline src/main/java/com/budgetapp/controllers/IncomeController.java]: Make same changes as requested in the expense controller.

**Review by colsen-neslo — CHANGES_REQUESTED** [review]

File: src/main/java/com/budgetapp/controllers/IncomeController.java (hunk 10 above)
Comment (Stanley-Maraba) [reply src/main/java/com/budgetapp/controllers/IncomeController.java]: the changes i was resolving them before i push code please check again

File: src/main/java/com/budgetapp/controllers/ExpensesController.java [hunk 11]
```diff
@@ -42,4 +42,14 @@ public ResponseEntity<Expense> saveExpenses(@RequestBody final Expense expense)
         final var createExpenses = expensesService.addExpenses(expense);
         return ResponseEntity.status(HttpStatus.CREATED).body(createExpenses);
     }
+
+    @GetMapping(path = "/{userId}", produces = MediaType.APPLICATION_JSON_VALUE)
+    private ResponseEntity<List<Expense>> findIncomesById(@PathVariable final Long userId) {
+        List<Expense> expense = expensesService.findExpensesByUserId(userId);
+        if (expense != null) {
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/controllers/ExpensesController.java]: expense should never be null, this check is redundant

File: src/main/java/com/budgetapp/repositories/IncomeRepository.java [hunk 12]
```diff
@@ -3,5 +3,8 @@
 import com.budgetapp.entities.Income;
 import org.springframework.data.jpa.repository.JpaRepository;
 
+import java.util.List;
+
 public interface IncomeRepository extends JpaRepository<Income, Long> {
+    List<Income> findByUserId(final Long userId);
```
Comment (JOlsen-Neslo) [inline src/main/java/com/budgetapp/repositories/IncomeRepository.java:9]: I don't think this will relate to the correct SQL query using Spring data. When finding a list the keywords are `findAll` - `findBy` returns an optional of a single record. So for your case it must be `findAllBy`

**Review by JOlsen-Neslo — CHANGES_REQUESTED** [review]

File: src/main/java/com/budgetapp/repositories/IncomeRepository.java (hunk 12 above)
Comment (Stanley-Maraba) [reply src/main/java/com/budgetapp/repositories/IncomeRepository.java:9]: @colsen-neslo @JOlsen-Neslo 

it relates or maybe i am not understanding.  

  List<Income> findByUserId(final Long userId);

Hibernate: select i1_0.id,i1_0.name,i1_0.user_id,i1_0.value from incomes i1_0 where i1_0.user_id=?

Comment (JOlsen-Neslo) [reply src/main/java/com/budgetapp/repositories/IncomeRepository.java:9]: The general pattern is `findAll` for a list, but if that works then its fine

**Review by JOlsen-Neslo — APPROVED** [review]

File: src/main/java/com/budgetapp/repositories/IncomeRepository.java (hunk 12 above)
Comment (Stanley-Maraba) [reply src/main/java/com/budgetapp/repositories/IncomeRepository.java:9]: ohh alright i get you now thanks

**Review by colsen-neslo — APPROVED** [review]
